libtenzir/builtins/operators/load_tcp.cpp

```diff
@@ -0,0 +1,965 @@
+    if (url.inner.starts_with("tcp://")) {
+      url.inner = url.inner.substr(6);
+      url.source.begin += 6;
+    }
+    if (const auto splits = detail::split(url.inner, ":", 1);
+        splits.size() != 2) {
+      diagnostic::error("malformed URL")
+        .primary(url.source)
+        .hint("syntax: tcp://<hostname>:<port>")
+        .usage(parser.usage())
+        .docs(parser.docs())
+        .emit(ctx);
+      failed = true;
+    } else {
+      args.endpoint.inner.hostname = std::string{splits[0]};
+      args.endpoint.inner.port = std::string{splits[1]};
+      args.endpoint.source = url.source;
+    }
+    if (parallel) {
+      args.parallel = *parallel;
+      if (args.parallel.inner == 0
+          or args.parallel.inner > std::thread::hardware_concurrency()) {
+        diagnostic::error("`parallel` must be between 1 and {}",
+                          std::thread::hardware_concurrency())
+          .primary(parallel->source)
+          .usage(parser.usage())
+          .docs(parser.docs())
+          .emit(ctx);
+        failed = true;
+      }
+      if (args.connect and args.parallel.inner != 1) {
+        diagnostic::warning("`parallel` is ignored when `connect` is set")
+          .primary(*args.connect)
+          .primary(parallel->source)
+          .usage(parser.usage())
+          .docs(parser.docs())
+          .emit(ctx);
+      }
+    } else {
+      args.parallel = {1, inv.self.get_location()};
+    }
+    if (tls and not tls->inner) {
+      if (args.certfile) {
+        diagnostic::error("conflicting option: `certfile` requires `tls`")
+          .primary(tls->source)
+          .primary(args.certfile->source)
+          .usage(parser.usage())
+          .docs(parser.docs())
+          .emit(ctx);
+        failed = true;
+      }
+      if (args.keyfile) {
+        diagnostic::error("conflicting option: `keyfile` requires `tls`")
+          .primary(tls->source)
+          .primary(args.keyfile->source)
+          .usage(parser.usage())
+          .docs(parser.docs())
+          .emit(ctx);
+        failed = true;
+      }
+    }
+    if (tls) {
+      args.tls.emplace(tls->source);
+    } else if (args.certfile or args.keyfile) {
+      args.tls.emplace(inv.self.get_location());
+    }
+    if (not args.pipeline) {
+      // If the user does not provide a pipeline, we fall back to just an empty
+      // pipeline, i.e., pass the bytes for all connections through.
+      args.pipeline.emplace(pipeline{}, inv.self.get_location());
+    }
+    const auto output_type = args.pipeline->inner.infer_type(tag_v<chunk_ptr>);
+    if (not output_type) {
+      diagnostic::error(output_type.error())
+        .note("failed to infer output type of nested pipeline")
+        .primary(args.pipeline->source)
+        .usage(parser.usage())
+        .docs(parser.docs())
+        .emit(ctx);
+      failed = true;
+    }
+    if (failed) {
+      return failure::promise();
+    }
+    return output_type->match(
```

Comment:
Should we also check here that the pipeline accepts bytes as input?

Reply:
This returns the output _iff_ the pipeline gets bytes as input (which it always does, as that's what the nested pipeline's source has as a fixed output). So this is working as expected.